src/repo.rs

```diff
@@ -1840,6 +1841,80 @@ impl Repository {
         Ok(())
     }
 
+    /// Merge two files as they exist in the in-memory data structures, using
+    /// the given common ancestor as the baseline, producing a
+    /// `git_merge_file_result` that reflects the merge result.  The
+    /// `git_merge_file_result` must be freed with `git_merge_file_result_free`.
+    ///
+    /// Note that this function does not reference a repository and any
+    /// configuration must be passed as `git_merge_file_options`.
+    ///
+    /// @param out The git_merge_file_result to be filled in
+    /// @param ancestor The contents of the ancestor file
+    /// @param ours The contents of the file in "our" side
+    /// @param theirs The contents of the file in "their" side
+    /// @param opts The merge file options or `NULL` for defaults
+    /// @return 0 on success or error code
+    pub fn merge_file(
+        &self,
+        ancestor: Option<&IndexEntry>,
+        ours: Option<&IndexEntry>,
+        theirs: Option<&IndexEntry>,
+        options: Option<&MergeFileOptions>,
+    ) -> Result<MergeFileResult, Error> {
+        let ancestor_input;
+        let ours_input;
+        let theirs_input;
+
+        let ancestor_raw;
+        let ours_raw;
+        let theirs_raw;
+
+        if let Some(ancestor) = ancestor {
+            ancestor_input = MergeFileInput::from(&self, ancestor);
+            ancestor_raw = ancestor_input.raw();
+        } else {
+            ancestor_raw = ptr::null();
+        }
+        if let Some(ours) = ours {
+            ours_input = MergeFileInput::from(&self, ours);
+            ours_raw = ours_input.raw();
+        } else {
+            ours_raw = ptr::null();
+        }
+        if let Some(theirs) = theirs {
+            theirs_input = MergeFileInput::from(&self, theirs);
+            theirs_raw = theirs_input.raw();
+        } else {
+            theirs_raw = ptr::null();
+        }
+
+        let mut ret = raw::git_merge_file_result {
+            automergeable: 0,
+            path: ptr::null(),
+            mode: 0,
+            ptr: ptr::null(),
+            len: 0,
+        };
+
+        unsafe {
+            try_call!(raw::git_merge_file(
+                &mut ret,
+                ancestor_raw,
+                ours_raw,
+                theirs_raw,
+                options.map(|o| o.raw())
+            ));
+
+            let result = MergeFileResult::from_raw(ret);
+
+            // FIXME: need to free????
+            raw::git_merge_file_result_free(&mut ret as *mut _);
```

Comment:
Typically this is modeled by `MergeFileResult` taking ownership of the `git_merge_file_result` and is then responsible for any deallocation necessary, would that be possible here?

Reply:
So `raw::git_merge_file_result_free(&mut ret as *mut _);` is not needed?